File: src/org/perses/util/TimeSpan.java
/*
 * Copyright (C) 2018-2020 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */

package org.perses.util;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Joiner;
import com.google.common.base.MoreObjects;

import java.io.Serializable;
import java.util.ArrayList;

import static com.google.common.base.Preconditions.checkArgument;

/** Representation of a time span. */
public final class TimeSpan implements Serializable {

  private final long startMillis;
  private final long endMillis;

  private TimeSpan(long startMillis, long endMillis) {
    checkArgument(startMillis <= endMillis);
    checkArgument(startMillis > 0);
    checkArgument(endMillis > 0);
    this.startMillis = startMillis;
    this.endMillis = endMillis;
  }

  public long getStartTimeMillis() {
    return startMillis;
  }

  public long getEndTimeMillis() {
    return endMillis;
  }

  public long getElapsedTimeInSeconds() {
    return getElapsedTimeInMillis() / 1000;
  }

  public long getElapsedTimeInMillis() {
    return getEndTimeMillis() - getStartTimeMillis();
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("startMillis", startMillis)
        .add("endMillis", endMillis)
        .add("formatted", getFormattedElaspedTime())
        .toString();
  }

  public String getFormattedElaspedTime() {
    long time = this.getElapsedTimeInSeconds();
    return formatSeconds(time);
  }

  public static final class Builder {

    private final long startMillis;

    @VisibleForTesting
    Builder(long startMillis) {
      this.startMillis = startMillis;
    }

    public static Builder start(long nowInMillis) {
      return new Builder(nowInMillis);
    }

    public TimeSpan end(long nowInMillis) {
      return createTimeSpan(nowInMillis);
    }

    public long getStartMillis() {
      return startMillis;
    }

    @VisibleForTesting
    TimeSpan createTimeSpan(long endMillis) {
      return new TimeSpan(startMillis, endMillis);
    }
  }

  private static String formatTimeComponent(long value, String component) {
    checkArgument(value > 0);
    if (value == 1) {
      return "1 " + component;
    } else {
      return value + " " + component + "s";
    }
  }

  private static String formatSeconds(long seconds) {
    if (seconds == 1) {
      return "1 second";
    }
    if (seconds <= 60) {
      return seconds + " seconds";
    }
    final StringBuilder builder = new StringBuilder();
    builder.append(seconds).append(" seconds (");

    ArrayList<String> components = new ArrayList<>(/*initialCapacity=*/ 5);

    final long days = seconds / (24 * 3600);
    if (days > 0) {
      components.add(formatTimeComponent(days, "day"));
    }
    seconds %= 24 * 3600;

    final long hours = seconds / 3600;
    if (hours > 0) {
      components.add(formatTimeComponent(hours, "hour"));
    }
    seconds %= 3600;

    final long minutes = seconds / 60;
    if (minutes > 0) {
      components.add(formatTimeComponent(minutes, "minute"));
    }
    seconds %= 60;

    if (seconds > 0) {
      components.add(formatTimeComponent(seconds, "second"));
    }

    builder.append(Joiner.on(' ').join(components));
    builder.append(")");
    return builder.toString();
  }
}

File: src/org/perses/util/TimeSpan.kt
/*
 * Copyright (C) 2018-2026 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.util

import com.google.common.base.MoreObjects
import com.google.common.base.Strings
import java.io.Serializable

/** Representation of a time span.  */
class TimeSpan(
  val startTimeMillis: Long,
  val endTimeMillis: Long,
) : Serializable {
  init {
    require(startTimeMillis <= endTimeMillis)
    require(startTimeMillis >= 0)
    require(endTimeMillis > 0)
  }

  val endTimeInSeconds: Long
    get() = endTimeMillis / 1000

  val elapsedTimeInMillis: Long
    get() = endTimeMillis - startTimeMillis

  val elapsedTimeInSeconds: Int
    get() = (elapsedTimeInMillis / 1000).toInt()

  val breakdown: TimeSpanBreakdown by lazy {
    var seconds = elapsedTimeInSeconds
    val days = seconds / (24 * 3600)
    seconds %= 24 * 3600
    val hours = seconds / 3600
    seconds %= 3600
    val minutes = seconds / 60
    seconds %= 60
    TimeSpanBreakdown(days, hours, minutes, seconds)
  }

  override fun toString(): String =
    MoreObjects
      .toStringHelper(this)
      .add("startMillis", startTimeMillis)
      .add("endMillis", endTimeMillis)
      .add("formatted", formattedElapsedTime)
      .toString()

  val formattedElapsedTime: String by lazy {
    if (elapsedTimeInSeconds == 1) {
      "1 second"
    } else if (elapsedTimeInSeconds <= 60) {
      "$elapsedTimeInSeconds seconds"
    } else {
      "$elapsedTimeInSeconds seconds (${breakdown.formattedString})"
    }
  }

  class Builder internal constructor(
    val startMillis: Long,
  ) {
    fun end(nowInMillis: Long): TimeSpan = createTimeSpan(nowInMillis)

    fun createTimeSpan(endMillis: Long): TimeSpan = TimeSpan(startMillis, endMillis)

    companion object {
      @JvmStatic
      fun start(nowInMillis: Long): Builder = Builder(nowInMillis)
    }
  }

  data class TimeSpanBreakdown(
    val days: Int,
    val hours: Int,
    val minutes: Int,
    val seconds: Int,
  ) {
    val formattedString: String by lazy {
      val components = mutableListOf<String>()
      addTimeComponentToListIfNecessary(days, "day", components)
      addTimeComponentToListIfNecessary(hours, "hour", components)
      addTimeComponentToListIfNecessary(minutes, "minute", components)
      addTimeComponentToListIfNecessary(seconds, "second", components)
      components.joinToString(separator = " ")
    }

    fun computeConciseFormattedString(includeSeconds: Boolean): String {
      val builder = StringBuilder()
      if (days > 0) {
        builder.append(days).append('D')
      }
      if (builder.isNotEmpty()) {
        builder.append(':')
      }
      builder.append(Strings.padStart(hours.toString(), 2, '0'))
      builder.append(":").append(Strings.padStart(minutes.toString(), 2, '0'))
      if (includeSeconds) {
        builder.append(":").append(Strings.padStart(seconds.toString(), 2, '0'))
      }
      return builder.toString()
    }

    private fun addTimeComponentToListIfNecessary(
      value: Int,
      singularName: String,
      resultCollector: MutableList<String>,
    ) {
      if (value == 1) {
        resultCollector.add("1 $singularName")
      } else if (value > 1) {
        resultCollector.add("$value ${singularName}s")
      }
    }
  }
}
